### PR TITLE
Nodejs week3

### PR DESCRIPTION
**Pull request by l0rians**

### CODE REVIEW

File: api/src/routers/meals.js [hunk 1]
```diff
@@ -3,16 +3,61 @@ import knex from "../database_client.js";
...
 router.get("/", async (req, res, next) => {
   try {
-    const meals = await knex.from("meal");
+    let query = knex("meal");
+    if (req.query.maxPrice) {
+      const maxPrice = parseInt(req.query.maxPrice);
+      query = query.where("price", "<=", maxPrice);
+    }
+
+    if (req.query.availableReservations) {
```
Comment (kralle333) [inline api/src/routers/meals.js:18]: and what if its false? :)

File: api/src/routers/meals.js [hunk 2]
```diff
@@ -3,16 +3,61 @@ import knex from "../database_client.js";
...
+      const available = req.query.availableReservations === "true";
+      query = query
+        .leftJoin("reservation", "meal.id", "reservation.meal_id")
+        .groupBy("meal.id")
+        .select("meal.id", "meal.title", "meal.price", "meal.max_reservations")
+        .having(
+          `meal.max_reservations - COUNT(reservation.id) ${available ? ">" : "="} 0`
```
Comment (kralle333) [inline api/src/routers/meals.js:25]: creative with the subtraction!

File: api/src/routers/meals.js [hunk 3]
```diff
@@ -3,16 +3,61 @@ import knex from "../database_client.js";
...
+
+    if (req.query.dateBefore) {
+      const dateBefore = req.query.dateBefore;
+      query = query.where("when", "<", dateBefore);
+    }
+
+    if (req.query.limit) {
+      const limit = parseInt(req.query.limit);
+      query = query.limit(limit);
+    }
```
Comment (kralle333) [inline api/src/routers/meals.js]: Looks good, but you are missing some validation of the request here :) Think about it from the users perspective. If you were using an endpoint, wouldn't you appreciate if the server told you what part of the query was wrong?

For example limit must be an integer, but we don't tell the user that the value you gave us is invalid.

Use error code 400 for these errors

File: api/src/routers/reviews.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+// router to get a specific review
+router.get("/meal/:meal_id", async (req, res, next) => {
+  const { meal_id } = req.params;
+  try {
+    const reviews = await knex("review").where({ meal_id });
+    res.json(reviews);
+  } catch (error) {
+    next(error);
+  }
+});
```
Comment (kralle333) [inline api/src/routers/reviews.js]: you are missing some logic here :) Also remember to handle the case where the meal simply isnt found. Hint: Use error code 404 for for not found!

Comment (kralle333) [reply api/src/routers/reviews.js]: Also what is the meal doing here? :smile:

File: api/src/routers/reviews.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+router.post("/", async (req, res, next) => {
+  try {
+    const newReview = req.body;
+    const [reviewId] = await knex("review").insert(newReview);
```
Comment (kralle333) [inline api/src/routers/reviews.js:36]: not too familar with knex, but does this really only return the id?

**Review by kralle333 — APPROVED** [review]
Great work!
I have some comments that would improve your code, which I think you should look into. 

Keep up the good work!

File: api/src/routers/reviews.js (hunk 5 above)
Comment (l0rians) [reply api/src/routers/reviews.js:36]: image

File: api/src/routers/reviews.js (hunk 4 above)
Comment (l0rians) [reply api/src/routers/reviews.js]: meal id is used to filter reviews related to a specific dish. its looking for all reviews for this specific meal id

**Pull request merged**